tests/cases/conformance/types/typeRelationships/typeInference/noInfer.ts

```diff
@@ -0,0 +1,78 @@
+// @strict: true
+// @declaration: true
+
+// NoInfer<T> is erased for primitives
+
+type T00 = NoInfer<string>;
+type T01 = NoInfer<string | number | boolean>;
+type T02 = NoInfer<undefined>;
+type T03 = NoInfer<"foo">;
+type T04 = NoInfer<`foo${string}`>;
+type T05 = NoInfer<`foo${string}` & `${string}bar`>;
+type T06 = NoInfer<{}>;
+
+// NoInfer<T> is preserved for object types
+
+type T10 = NoInfer<string[]>;
+type T11 = NoInfer<{ x: string }>;

```

Comment:
Any particular reason why those can't be erased?

At the moment this, in part at least, creates problems with mapped types and `keyof`. This works OK:
```ts
declare const obj: NoInfer<{ a: string, b: number }>
const ok: keyof typeof obj = 'a'
const err: keyof typeof obj = 'c'
```

However, `shouldDeferIndexType` returns `true` for this type and I feel like it shouldn't - I'd expect that this might create some problems.

In a similar vein this isn't reduced eagerly:
```ts
declare const obj: NoInfer<{ a: string, b: number }>
type Mapped/**/ = { [K in keyof typeof obj]: K }
```

If we hover over `Mapped` we'll see the mapped type instead of its evaluated form (`isGenericMappedType` return `true` since `isGenericIndexType(constraint)` is `true`).

I didn't analyze this to its full extent but I think that it's very likely that this has a direct impact on:
```ts
class Animal {
  move() {}
}
class Dog extends Animal {
  woof() {}
}
class Cat extends Animal {
  meow() {}
}

type K<T> = T extends any ? { [k in keyof T]: string } : never;

export type UnionToIntersection<U> = (
  U extends any ? (k: U) => void : never
) extends (k: infer I) => void
  ? I
  : never;

function doSomething<T extends Dog | Cat>(
  getDefault: (x: any) => T,
  value: UnionToIntersection<K<NoInfer<T>>>,
) {}

doSomething((x) => new Dog(), { woof: "", move: "" });
doSomething(() => new Dog(), { woof: "", move: "" });
```

Both calls fail here while the second one succeeds without `NoInfer`. I think that both should succeed with `NoInfer` though. At the moment, the first call infers `Dog | Cat` and the second one infers `Dog` but they both should infer `Dog`

Reply:
> Any particular reason why those can't be erased?

We don't have infrastructure to consistently determine if an object type contains generics. It would require deep traversal, detection of recursion, and so on. Basically the equivalent of the traversal we do in `inferTypes`. We could consider using `couldContainTypeVariables` to guide erasure of `NoInfer<T>`, but those are optimization heuristics that all of a sudden would become observable.

> At the moment this, in part at least, creates problems with mapped types and `keyof`.

Agreed, we should erase `NoInfer<T>` when it is the argument of `keyof`. I'll make that change.

Reply:
The weird case from the above works now 🎉 [TS playground](https://www.staging-typescript.org/play?ts=5.4.0-pr-56794-29#code/MYGwhgzhAECCB2BLAtmE0DeBYAUNayA9gG4CmAFAJSYC+udOokMAIoQObSkAeALqfAAmMBCjSZc+AO6FCAMyq16uJlGgBhMLy58BwuElTpseAqUJTFGBg1y8AngAdS0ANIAeACoA+aAF5oTx1+IRgweHtoAH5MaABtAGtoRHhoBNJ7eUCAXQAuaAheACcUzhpofPhSMiKAblxcHkdCIu0HZ2gAVSRCeE9CAEl4fiKIUmBeRF73Tt8A8kku4L0wiOjocgT8zuo-X2JCREEK6Cqa3GoeEP1N-JS5UiLoAd39w8FFmIHFyurH+pwuDkAFd4BMpqlBIQAMqEZCkXgAC1KXmWoWgbE4AB8NFpvAtTOwESxSHIwMCQLx8uRuPlwvZXoEADSLYhoYGkbY9PqDYaPMbg6YeAByvIeRS83ilLJw1GsDRwUNh8KRpXINMZVSkGI4VCZsRk8nyACJjfqiGQTcboDRKAClXCEcj4Ox1ZrSNrMXqDbI5FbzSROdBTTa7bggA). 

> Agreed, we should erase NoInfer<T> when it is the argument of keyof

Isn't this erasure a little bit too eager now? [TS playground](https://www.staging-typescript.org/play?ts=5.4.0-pr-56794-29&ssl=7&ssc=9&pln=1&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMBGAHgBV4QAPQ1YI+Ab3igH4AueUgGngAjbvABM8AL4A+ABSwA5jwDWIAJ45E8SgEoelALAAoMHhLw4RZBDLwAvAWJkZAcijPtRgPSf4v+AD0OIyNQSFgEFHRsPAcSUSoaehBGZjZOHn4hEXFpORhFeBV1TQA5HABJVEQQGCopXS0jE1QzCysMcXtCOJc3D0NvPwCgwyA)

```ts
declare function test1<T extends { a?: 1, b?: 2 }>(arg: keyof T): T
const result1 = test1('a')
//    ^? const result1: { a: any }

declare function test2<T extends { a?: 1, b?: 2 }>(arg: keyof NoInfer<T>): T
const result2 = test2('a')
//    ^? const result2: { a: any }
```

If we use `NoInfer<keyof T>` then it falls back to the constraint - shouldn't the same happen with `test2`? I see that `keyof` part here isn't wrapped with `NoInfer` but `T` is wrapped and `T` is the target of the inference here. So this creates a weird precedence in which a wrapped inference target is still inferred.

Reply:
inferring `T` through `keyof T` is already a weird-ish thing so perhaps this is a better demo of the problem with reverse mapped types:
```ts
declare function test1<T>(
  o1: {
    [K in keyof T]?: [T[K]];
  },
  o2: {
    [K in keyof T]: T[K];
  },
): T;
const result1 = test1({ a: [1] }, { a: 1, b: 2 });
//    ^?

declare function test2<T>(
  o1: {
    [K in keyof NoInfer<T>]?: [T[K]];
  },
  o2: {
    [K in keyof T]: T[K];
  },
): T;
const result2 = test2({ a: [1] }, { a: 1, b: 2 });
```

I'd expect the second call to succeed here. Wrapping `T` directly with `NoInfer` here is more intuitive than having to wrap whole `keyof T` - solely because `T` is the inference target and not `keyof T`.

Reply:
Yeah, erasure was a bit too aggressive. I think the proper solution is to transform `keyof NoInfer<T>` into `NoInfer<keyof T>`. The latest commits implement that.

Reply:
An interesting solution to the problem! I think it might be worth adding some of the test cases I mentioned here since based on the emit/display alone it's not quite apparent **why** such transformation happens and what could go wrong if that would be missing. I **very often** disable some parts of the compiler to learn what tests break without them to learn how things work and what I might have to support in my PRs etc. Tests like that helps me quite a bit.

For convenience, I'm putting all of them in the details here - feel free to pick up/ignore any of them:

<details>
<summary>Potential test cases</summary>

```ts
class Animal { move() {} }
class Dog extends Animal { woof() {} }
class Cat extends Animal { meow() {} }

type K<T> = T extends any ? { [k in keyof T]: string } : never;
type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never;

function doSomething<T extends Dog | Cat>(getDefault: (x: any) => T, value: UnionToIntersection<K<NoInfer<T>>>) {}

doSomething((x) => new Dog(), { woof: "", move: "" });
doSomething(() => new Dog(), { woof: "", move: "" });

declare function test1<T extends { a?: 1, b?: 2 }>(arg: keyof NoInfer<T>): T
const result1 = test1('a')

declare function test2<T>(
  o1: {
    [K in keyof NoInfer<T>]?: [T[K]];
  },
  o2: {
    [K in keyof T]: T[K];
  },
): T;
const result2 = test2({ a: [1] }, { a: 1, b: 2 });
```
</details>